otherlibs/systhreads/st_stubs.c

```diff
@@ -385,211 +308,326 @@ static value caml_thread_new_descriptor(value clos)
 static void caml_thread_remove_info(caml_thread_t th)
 {
   if (th->next == th)
-    all_threads = NULL; /* last OCaml thread exiting */
-  else if (all_threads == th)
-    all_threads = th->next;     /* PR#5295 */
+    All_threads = NULL; /* last OCaml thread exiting */
+  else if (All_threads == th)
+    All_threads = th->next;     /* PR#5295 */
   th->next->prev = th->prev;
   th->prev->next = th->next;
-#ifndef NATIVE_CODE
-  caml_stat_free(th->stack_low);
-#endif
-  if (th->backtrace_buffer != NULL) caml_stat_free(th->backtrace_buffer);
+  caml_free_stack(th->current_stack);
   caml_stat_free(th);
+  return;
 }
 
 /* Reinitialize the thread machinery after a fork() (PR#4577) */
+/* TODO(engil): more work on the multicore fork machinery. */
 
 static void caml_thread_reinitialize(void)
 {
-  struct channel * chan;
+  caml_thread_t th, next;
 
-  /* Remove all other threads (now nonexistent)
-     from the doubly-linked list of threads */
-  while (curr_thread->next != curr_thread) {
-    caml_memprof_delete_th_ctx(curr_thread->next->memprof_ctx);
-    caml_thread_remove_info(curr_thread->next);
+  th = Current_thread->next;
+  while (th != Current_thread) {
+    next = th->next;
+    caml_free_stack(th->current_stack);
+    caml_stat_free(th);
+    th = next;
   }
-
-  /* Reinitialize the master lock machinery,
-     just in case the fork happened while other threads were doing
-     caml_leave_blocking_section */
-  st_masterlock_init(&caml_master_lock);
-  /* Tick thread is not currently running in child process, will be
-     re-created at next Thread.create */
-  caml_tick_thread_running = 0;
-  /* Destroy all IO mutexes; will be reinitialized on demand */
-  for (chan = caml_all_opened_channels;
-       chan != NULL;
-       chan = chan->next) {
-    if (chan->mutex != NULL) {
-      st_mutex_destroy(chan->mutex);
-      chan->mutex = NULL;
+  Current_thread->next = Current_thread;
+  Current_thread->prev = Current_thread;
+  All_threads = Current_thread;
+
+  // within the child, the domain_lock needs to be reset
+  // and acquired.
+  caml_reset_domain_lock();
+  caml_acquire_domain_lock();
+  // master_lock needs to be initialized again.
+  // this process will also be the effective owner of the lock.
+  // so there is no need to run st_masterlock_acquire (busy = 1)
+  st_masterlock_init(&Thread_main_lock);
+}
+
+CAMLprim value caml_thread_initialize(value unit);
+CAMLprim value caml_thread_initialize_domain(value unit);
```

Comment:
I don't think you need  these forward declarations.  However, if you remove them, `caml_thread_initialize_domain` should be marked  `CAMLprim` when it is defined.